pkg/yqlib/operator_divide.go

```diff
@@ -0,0 +1,88 @@
+package yqlib
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	yaml "gopkg.in/yaml.v3"
+)
+
+func createDivideOp(lhs *ExpressionNode, rhs *ExpressionNode) *ExpressionNode {
+	return &ExpressionNode{Operation: &Operation{OperationType: divideOpType},
+		LHS: lhs,
+		RHS: rhs}
+}
+
+// func divideAssignOperator(d *dataTreeNavigator, context Context, expressionNode *ExpressionNode) (Context, error) {
+// 	return compoundAssignFunction(d, context, expressionNode, createDivideOp)
+// }
+
+func divideOperator(d *dataTreeNavigator, context Context, expressionNode *ExpressionNode) (Context, error) {
+	log.Debugf("Divide operator")
+
+	return crossFunction(d, context.ReadOnlyClone(), expressionNode, divide, false)
+}
+
+func divide(d *dataTreeNavigator, context Context, lhs *CandidateNode, rhs *CandidateNode) (*CandidateNode, error) {
+	lhs.Node = unwrapDoc(lhs.Node)
+	rhs.Node = unwrapDoc(rhs.Node)
+
+	lhsNode := lhs.Node
+
+	if lhsNode.Tag == "!!null" {
+		return nil, fmt.Errorf("%v (%v) cannot be divided by %v (%v)", lhsNode.Tag, lhs.GetNicePath(), rhs.Node.Tag, rhs.GetNicePath())
+	}
+
+	target := lhs.CreateReplacement(&yaml.Node{
+		Anchor: lhs.Node.Anchor,
+	})
+
+	if lhsNode.Kind == yaml.ScalarNode && rhs.Node.Kind == yaml.ScalarNode {
+		if err := divideScalars(context, target, lhsNode, rhs.Node); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("%v (%v) cannot be divided by %v (%v)", lhsNode.Tag, lhs.GetNicePath(), rhs.Node.Tag, rhs.GetNicePath())
+	}
+
+	return target, nil
+}
+
+func divideScalars(context Context, target *CandidateNode, lhs *yaml.Node, rhs *yaml.Node) error {
+	lhsTag := lhs.Tag
+	rhsTag := guessTagFromCustomType(rhs)
+	lhsIsCustom := false
+	if !strings.HasPrefix(lhsTag, "!!") {
+		// custom tag - we have to have a guess
+		lhsTag = guessTagFromCustomType(lhs)
+		lhsIsCustom = true
+	}
+
+	if lhsTag == "!!str" && rhsTag == "!!str" {
+		target.Node = split(lhs.Value, rhs.Value)
+		target.Node.Anchor = lhs.Anchor
+	} else if (lhsTag == "!!int" || lhsTag == "!!float") && (rhsTag == "!!int" || rhsTag == "!!float") {
+		target.Node.Kind = yaml.ScalarNode
+		target.Node.Style = lhs.Style
+
+		lhsNum, err := strconv.ParseFloat(lhs.Value, 64)
+		if err != nil {
+			return err
+		}
+		rhsNum, err := strconv.ParseFloat(rhs.Value, 64)
+		if err != nil {
+			return err
+		}
+		quotient := lhsNum / rhsNum
```

Comment:
what happens if you divide by zero?

Reply:
(be good to have a test so its in the docs)

Reply:
```
./yq -n '1 / 0'
+Inf
```

Reply:
https://github.com/mikefarah/yq/pull/1593/commits/16997769e0146cf285e9a75a5e7c27300fbb7ffb

Reply:
I was wondering this morning if +Inf is supported in json, and I guess it is not. NaN is the same.  What's your opinion on this?

https://stackoverflow.com/questions/1423081/json-left-out-infinity-and-nan-json-status-in-ecmascript



```bash
./yq -o json -n '{"a": 1 / 0}'
Error: yaml: cannot decode !!str `+Inf` as a !!float

jq -n '{"a": +Inf}'
jq: error: syntax error, unexpected '+' (Unix shell quoting issues?) at <top-level>, line 1:
{"a": +Inf}
jq: 1 compile error

 jq -n '{"a": infinite}'
{
  "a": 1.7976931348623157e+308
}
```

https://stedolan.github.io/jq/manual/

>  infinite, nan, isinfinite, isnan, isfinite, isnormal
> 
> Some arithmetic operations can yield infinities and "not a number" (NaN) values. The isinfinite builtin returns true if its input is infinite. The isnan builtin returns true if its input is a NaN. The infinite builtin returns a positive infinite value. The nan builtin returns a NaN. The isnormal builtin returns true if its input is a normal number.
> 
> Note that division by zero raises an error.
> 
> Currently most arithmetic operations operating on infinities, NaNs, and sub-normals do not raise errors.